File: balanced-parentheses/balancedParentheses.py
from ArrayStack import ArrayStack


def is_matched(expr):
    """ Returns true if the arithmatic express is true, false otherwise """
    left = '({['
    right = ')}]'
    s = ArrayStack()
    for c in expr:
        if c in left:
            s.push(c)
        elif c in right:
            if s.is_empty():
                return False
            if right.index(c) != left.index(s.pop()):
                return False
    return s.is_empty()


def main():
    expr = '[(3-4)+(x+6)]'
    print(is_matched(expr))


if __name__ == '__main__':
    main()

File: balanced-parentheses/balancedParentheses.js
/**
 * Write a function that takes an arithmatic expression
 * as parameter and returns true on balanced parentheses,
 * false otherwise.
 * 
 * Example: '[(3-4)+(x+6)]', '[{(a+b) * (c/d)}]' etc.
 * 
 * @param {String} expr 
 * 
 * @return {Boolean}
 */

function is_matched(expr) {
  const left = '({[';
  const right = ')}]';
  const a = [];

  for (let c of expr) {
    if (left.indexOf(c) !== -1) {
      a.push(c);
    } else if (right.indexOf(c) !== -1) {
      if (a.length === 0) {
        return false;
      }

      if (right.indexOf(c) !== left.indexOf(a.pop())) {
        return false;
      }
    }
  }

  return a.length === 0;
}

const expr = '[(3-4)+(x+6)]';
console.log(is_matched(expr));